### PR TITLE
chore: move s3 content to lib

### PR DESCRIPTION
**Pull request by gruyaume**
# Description

We take the content of the `s3` helpers file from the Vault machine charm, move it to a charm library and refactor the s3 related content in the charm to take advantage of it. 

As a follow-up PR we will want to fetch this lib in the machine charm. There, the change will be much smaller as the charm code will be identical from its current state.

## Rationale

The s3 logic is the same in both the machine and k8s charm and for the same reason that we have a Vault client library, we also have one for s3. The changes in the s3 lib were outlined in a previous PR in the vault machine charm. In summary:
- We don't require the charm to know about the lib's dependency errors (ex. botocore errors) and the only errors that bubble up are custom `S3Error`. This reduces the amount of information the charm code needs to know about the implementation of the s3 lib. 
- We add type checking using `boto3-stubs[s3]`

# Checklist:

- [ ] My code follows the [style guidelines](/CONTRIBUTING.md) of this project
- [ ] I have performed a self-review of my own code
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that validate the behaviour of the software
- [ ] I validated that new and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

**Review by saltiyazan — CHANGES_REQUESTED** [review]
Beautiful, I added a few minor comments.
One general suggestion I have, and we can include it in this PR or not is to move these two S3 related functions from the charm file to the lib:
`_get_missing_s3_parameters`
`_get_s3_parameters`

I don't think we are using or doing anything strictly related to the charm here.

**Comment by gruyaume** [comment]
> Beautiful, I added a few minor comments. One general suggestion I have, and we can include it in this PR or not is to move these two S3 related functions from the charm file to the lib: `_get_missing_s3_parameters` `_get_s3_parameters`
> 
> I don't think we are using or doing anything strictly related to the charm here.

I would leave those out as they interact with the relation data (through `s3_requirer`) and are not pure s3 calls.

File: tests/unit/lib/charms/vault_k8s/v0/test_vault_s3.py
[the remark this answers is not in the document]
Comment (gruyaume) [reply tests/unit/lib/charms/vault_k8s/v0/test_vault_s3.py:1]: Done

File: src/charm.py
[the remark this answers is not in the document]
Comment (gruyaume) [reply src/charm.py:1]: Done

**Review by saltiyazan — APPROVED** [review]

File: src/charm.py [hunk 1]
```diff
@@ -568,18 +567,12 @@ def _on_create_backup_action(self, event: ActionEvent) -> None:
...
-            event.fail(message="Only leader unit can perform backup operations.")
-            return
-
-        s3_requirements, error_message = self._check_s3_requirements()
-        if not s3_requirements:
-            logger.error(error_message)
-            event.fail(message=f"{error_message} Failed to perform backup.")
+        s3_pre_requisites_err = self._check_s3_pre_requisites()
+        if s3_pre_requisites_err:
+            event.fail(message=f"S3 pre-requisites not met. {s3_pre_requisites_err}.")
```
[the remark this answers is not in the document]
Comment (gruyaume) [reply src/charm.py:572]: Indeed, I couldn't find a better way to do this

File: lib/charms/vault_k8s/v0/vault_s3.py [hunk 2]
```diff
@@ -57,42 +87,48 @@ def create_bucket(self, bucket_name: str) -> bool:
...
+        """Return whether the bucket exists."""
+        try:
+            bucket.meta.client.head_bucket(Bucket=bucket.name)
         except ClientError:
-            logger.info("Bucket %s doesn't exist, creating it.", bucket_name)
-            pass
-        except BotoCoreError as e:
-            logger.error("Failed to check whether bucket exists. %s", e)
             return False
+        except BotoCoreError:
```
[the remark this answers is not in the document]
Comment (gruyaume) [reply lib/charms/vault_k8s/v0/vault_s3.py]: done

**Review by Kayra1 — APPROVED** [review]

File: src/charm.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (Kayra1) [reply src/charm.py:572]: its the best way

**Pull request merged**